### ISSUE TITLE
[Documentation/Example wrong] reference dataset in lgb.Dataset and early stopping

### ISSUE THREAD

**Issue by dancaspi** [Microsoft/LightGBM#136]
In the basic example:
https://github.com/Microsoft/LightGBM/blob/master/examples/python-guide/simple_example.py

the final number of trees is determined by the result of early stopping. is the test set only used for evaluation purposes (show the result) or actually used in determining the number of trees? as if it's latter, this poses a statistical question: the number of trees is just like any other hyper-parameter (any other views on this?), if so, why would you fit this parameter on the test and not on the train? if this is not the case, and the number of trees are fitted solely on the train,what is the use of the test set in the evaluation? just to show you the performance?  we are "predicting" the test set later in the code anyway...

while this isn't a code question, i think this sample will be used as base code for many applications, and should be statistically accurate, clearly i may be wrong from a statistical point of view, but would love to get an answer...

I really hope I didn't say complete nonsense above, but worth the risk... 

Dan

**Comment by gugatr0n1c** [comment]
Idea of early_stopping is to avoid overfitting and stop training when model start to overfit.
If you will wait until train_data start to overfit/diverge, than it is often too late.
Yes you can watch on num_trees as a next hyperparameter, but actually hyperparametrs are difficult to fit on train_data (there are some concept on neural nets where architecture is changing during learning process).

So in ideal case you have 3 groups of data:
train_data - for building trees - fitting directly
valid_data - for optimizing hyperparameters (and so for early_stopping)
test_data - independent set, where you can test performance

You can try to to early stopping with train_data, you will have much better performance on train_data, but it is only heavily overfitted. When you test it against valid or test data you will see difference.

**Comment by wxchan** [comment]
The example is meant to let user know the **'syntax'** only. In reality, eval set and test set are different.

**Comment by dancaspi** [comment]
@gugatr0n1c  - I am not sure that the separation you make between "building trees" and "hyperparameters" is statistically justified - from a statistical standpoint, at least to me, they are all fitted on data which is never used as "test". I do agree that you can decide (for whichever reason), to tune some of the parameters on the validation set, without harm the validity of your evaluation on the test set. From a practical standpoint, this means you have to split your train into train/validation - but then one must consider if the implications of reducing train, and the required validation size. however, **i believe this is going off topic here, and that we are overall in agreement**

@wxchan I agree completely, and this was my original intention, this demonstrates how to use lightgbm, and will be used ubiquitously by future users, therefore I believe it's worth changing the example to include a validation set (maybe split the train). just a suggestion... 

thank you both of you for your replies,

Dan

**Comment by gugatr0n1c** [comment]
@dancaspi oh sorry Dan, I do not understand your original point. Yes, agree fully. There is still test sets out there to tell the results (measure, prediction,..), but train or valid is still about fitting data - no matter how.

I am not sure if there is way how to tell how many iterations are optimal without some data_set to tell (in this case valid_set). We can guess, but I believe that ie. reduce train_data about 10 percent for valid set and then stop training up on this sets is better way. I know that reducing train_sets is very very painful, but what else we can do without 'cheating'? Sure there could be experiment:
1] use 80% on train and stop training up on this sets and then measure on test_sets
2] use 70% on train, 10% on valid, stop training on valid_set and then measure on test_test
if more data is better then valid set to stop..
Or do you have some idea?

**Comment by gugatr0n1c** [comment]
3] maybe run 1] find out optimum, reduce ie 15% (expecting overfitting)

**Comment by dancaspi** [comment]
@gugatr0n1c hey, thanks for your reply.
As I said, we are in agreement as we both say that fitting the number of trees (iterations) on the test is flawed. Now remains the question of "early stopping", and is it at all meaningful on the train set. as the train set error is "always decreasing", we could argue that using early stopping on **train** is meaningless, and thus the mechanism of early stopping is only relevant with a third (validation) set. On the other hand, you could just use simple train/test, have a "reasonable" number of iterations in your grid, and consider the number of iterations just like any other hyper-parameter, tuned via cross validation on the train set. both approaches seem valid (to my truly humble opinion), but each has its drawback (early stopping probably more accurate, but you pay with less data for the tuning of other parameters). 
what do you think? 

*However*, My key point in opening this post was that fitting **any** parameter using the test set is methodologically flawed, and since this is a "basic example", many future users of LightGBM would consider this as code to start with, and could assume that this is a valid methodology... 

I hope the above made sense,

Dan

**Comment by gugatr0n1c** [comment]
@dancaspi sure, I agree with test_set vs fitting parameters, that is clear..
